--- a/drivers/adc/core.c
+++ b/drivers/adc/core.c
@@ -8,6 +8,7 @@
 #include "adc_priv.h"
 
 struct list_head adc_host_head;
+struct adc_host *g_adc = NULL;
 
 struct adc_host *adc_alloc_host(struct device *dev, int extra, enum host_chn_mask mask)
 {
@@ -238,6 +239,34 @@
 
 EXPORT_SYMBOL(adc_async_read);
 
+int adc_get_def_ref_volt(void)
+{
+        return g_adc->pdata->ref_volt;
+}
+EXPORT_SYMBOL(adc_get_def_ref_volt);
+
+int adc_get_curr_ref_volt(void)
+{
+        int v = 0, volt = 0;
+
+        if(!g_adc)
+                return -EINVAL;
+        if(!g_adc->base_client)
+                return g_adc->pdata->ref_volt;
+
+        volt = g_adc->pdata->get_base_volt();
+        if(volt < 0)
+                return g_adc->pdata->ref_volt;
+        
+        v = adc_sync_read(g_adc->base_client);
+        if(v < 0)
+                return v;
+
+        return volt * 1024 / v;
+}
+EXPORT_SYMBOL(adc_get_curr_ref_volt);
+
+
 static int __init adc_core_init(void)
 {
         INIT_LIST_HEAD(&adc_host_head);
