--- a/drivers/mmc/host/rk29_sdmmc.c
+++ b/drivers/mmc/host/rk29_sdmmc.c
@@ -73,7 +73,7 @@
 #if defined(CONFIG_ARCH_RK29)
     #define RK29_SDMMC0DETECTN_GPIO RK29_PIN2_PA2
     #define RK29_SDMMC0PWREN_GPIO   RK29_PIN5_PD5
-#elif defined(CONFIG_ARCH_RK3066B)
+#elif defined(CONFIG_ARCH_RK3066B)||defined(CONFIG_ARCH_RK3188)
     #define RK29_SDMMC0DETECTN_GPIO RK30_PIN3_PB0
     #define RK29_SDMMC0PWREN_GPIO   RK30_PIN3_PA1
 #elif defined(CONFIG_ARCH_RK30)
