--- a/drivers/usb/storage/initializers.h
+++ b/drivers/usb/storage/initializers.h
@@ -46,5 +46,12 @@
  * flash reader */
 int usb_stor_ucr61s2b_init(struct us_data *us);
 
+/* This places the HUAWEI usb dongles in multi-port mode */
+int usb_stor_huawei_init(struct us_data *us);
+
 /* This places the HUAWEI E220 devices in multi-port mode */
 int usb_stor_huawei_e220_init(struct us_data *us);
+
+int usb_stor_zte_k4505_init(struct us_data *us);
+
+int usb_stor_zte_init(struct us_data *us);
