--- a/include/net/ipv6.h
+++ b/include/net/ipv6.h
@@ -660,4 +660,14 @@
 extern void ipv6_static_sysctl_unregister(void);
 #endif
 
+int icmpv6_push_pending_frames(struct sock *sk, struct flowi6 *fl6,
+			       struct icmp6hdr *thdr, int len);
+
+static inline bool __ipv6_addr_needs_scope_id(int type)
+{
+	return type & IPV6_ADDR_LINKLOCAL ||
+	       (type & IPV6_ADDR_MULTICAST &&
+		(type & (IPV6_ADDR_LOOPBACK|IPV6_ADDR_LINKLOCAL)));
+}
+
 #endif /* _NET_IPV6_H */
