--- a/kernel/time/alarmtimer.c
+++ b/kernel/time/alarmtimer.c
@@ -298,7 +298,7 @@
  * @type: the type of the alarm
  * @function: callback that is run when the alarm fires
  */
-void alarm_init(struct alarm *alarm, enum alarmtimer_type type,
+static void alarm_init(struct alarm *alarm, enum alarmtimer_type type,
 		void (*function)(struct alarm *))
 {
 	timerqueue_init(&alarm->node);
@@ -333,7 +333,7 @@
  * alarm_cancel - Tries to cancel an alarm timer
  * @alarm: ptr to alarm to be canceled
  */
-void alarm_cancel(struct alarm *alarm)
+static void alarm_cancel(struct alarm *alarm)
 {
 	struct alarm_base *base = &alarm_bases[alarm->type];
 	unsigned long flags;
