--- a/arch/arm/plat-rk/include/plat/board.h
+++ b/arch/arm/plat-rk/include/plat/board.h
@@ -147,6 +147,12 @@
     struct rksdmmc_gpio   data1_gpio;    
     struct rksdmmc_gpio   data2_gpio;
     struct rksdmmc_gpio   data3_gpio;
+#define USE_SDMMC_DATA4_DATA7  /*In order to be compatible with old project, which have not define the member used for eMMC */    
+    struct rksdmmc_gpio   data4_gpio;
+    struct rksdmmc_gpio   data5_gpio;    
+    struct rksdmmc_gpio   data6_gpio;
+    struct rksdmmc_gpio   data7_gpio;
+    struct rksdmmc_gpio   rstnout_gpio;
    
     struct rksdmmc_gpio   detect_irq;    
     struct rksdmmc_gpio   power_en_gpio;   
@@ -176,6 +182,7 @@
 	int (*io_init)(void);
 	int (*io_deinit)(void);
 	void (*set_iomux)(int device_id, unsigned int bus_width);//added by xbw at 2011-10-13
+	int (*emmc_is_selected)(int device_id);
 	int (*status)(struct device *);
 	int (*register_status_notify)(void (*callback)(int card_present, void *dev_id), void *dev_id);
 	int detect_irq;
@@ -186,6 +193,8 @@
 	int write_prt;
 	int write_prt_enalbe_level;
 	unsigned int sdio_INT_gpio; 
+	int sdio_INT_level;
+#define USE_SDIO_INT_LEVEL  /*In order to be compatible with old project, those who do not define the member  sdio_INT_level */
 	struct rksdmmc_gpio   det_pin_info;
         int (*sd_vcc_reset)(void);
 };
