diff --git a/drivers/mtd/rknand/rknand_base.h b/drivers/mtd/rknand/rknand_base.h
index b7fa95ab..c8274a91 100755
--- a/drivers/mtd/rknand/rknand_base.h
+++ b/drivers/mtd/rknand/rknand_base.h
@@ -72,7 +72,10 @@ struct rknand_info {
     void (*nand_timing_config)(unsigned long AHBnKHz);
     void (*rknand_suspend)(void);
     void (*rknand_resume)(void);
-    int reserved[20];
+    int reserved0;
+    int (*dev_cache_flush)(void);
+    int (*diacard)(int);
+    int reserved[17];
 };
 
 extern int rknand_queue_read(int Index, int nSec, void *buf);
