--- a/arch/arm/mach-rk3188/dvfs.c
+++ b/arch/arm/mach-rk3188/dvfs.c
@@ -26,6 +26,7 @@
 #include <linux/delay.h>
 #include <linux/io.h>
 #include <linux/hrtimer.h>
+#include <plat/efuse.h>
 
 static int rk_dvfs_clk_notifier_event(struct notifier_block *this,
 		unsigned long event, void *ptr)
@@ -77,6 +78,118 @@
 	.notifier_call = rk_dvfs_clk_notifier_event,
 };
 
+struct lkg_maxvolt {
+	int leakage_level;
+	unsigned int maxvolt;
+};
+#if 0
+/* avdd_com & vdd_arm separate circuit */
+static struct lkg_maxvolt lkg_volt_table[] = {
+	{.leakage_level = 1,	.maxvolt = 1350 * 1000},
+	{.leakage_level = 3,	.maxvolt = 1275 * 1000},
+	{.leakage_level = 15,	.maxvolt = 1200 * 1000},
+};
+#else
+/* avdd_com & vdd_arm short circuit */
+static struct lkg_maxvolt lkg_volt_table[] = {
+	{.leakage_level = 9000,		.maxvolt = 1350 * 1000},
+	{.leakage_level = 15000,	.maxvolt = 1300 * 1000},
+	{.leakage_level = 55000,	.maxvolt = 1250 * 1000},
+};
+#endif
+#define LOW_LEAKAGE_BOUND	2
+#define VOLT_COMPENSATION	(25000)	// uV
+static int leakage_level = 0;
+#define MHZ	(1000 * 1000)
+#define KHZ	(1000)
+// Delayline bound for nandc = 148.5MHz, Varm = Vlog = 1.00V
+#define HIGH_DELAYLINE	125
+#define LOW_DELAYLINE	125
+static u8 rk30_get_avs_val(void);
+void dvfs_adjust_table_lmtvolt(struct clk *clk, struct cpufreq_frequency_table *table)
+{
+	int i = 0;
+	unsigned int maxvolt = 0;
+	if (IS_ERR_OR_NULL(clk) || IS_ERR_OR_NULL(table)) {
+		DVFS_ERR("%s: clk error OR table error\n", __func__);
+		return ;
+	}
+
+	leakage_level = rk_leakage_val();
+	printk("DVFS MSG: %s: %s get leakage_level = %d\n", clk->name, __func__, leakage_level);
+	if (leakage_level == 0) {
+		
+		/*
+		 * This is for delayline auto scale voltage, 
+		 * FIXME: HIGH_DELAYLINE / LOW_DELAYLINE value maybe redefined under
+		 * Varm = Vlog = 1.00V.
+		 * Warning: this value is frequency/voltage sensitive, care
+		 * about Freq nandc/Volt log.
+		 *
+		 */
+		return;
+
+		unsigned long delayline_val = 0;
+		unsigned long high_delayline = 0, low_delayline = 0;
+		unsigned long rate_nandc = 0;
+		rate_nandc = clk_get_rate(clk_get(NULL, "nandc")) / KHZ;
+		printk("Get nandc rate = %lu KHz\n", rate_nandc);
+		high_delayline = HIGH_DELAYLINE * 148500 / rate_nandc;
+		low_delayline = LOW_DELAYLINE * 148500 / rate_nandc;
+		delayline_val = rk30_get_avs_val();
+		printk("This chip no leakage msg, use delayline instead, val = %lu.(HDL=%lu, LDL=%lu)\n",
+				delayline_val, high_delayline, low_delayline);
+
+		if (delayline_val >= high_delayline) {
+			leakage_level = 4;	//same as leakage_level > 4
+
+		} else if (delayline_val <= low_delayline) {
+			leakage_level = 1;
+			printk("Delayline TOO LOW, high voltage request\n");
+
+		} else
+			leakage_level = 2;	//same as leakage_level = 3
+	}
+
+	for (i = 0; i < ARRAY_SIZE(lkg_volt_table); i++) {
+		if (leakage_level <= lkg_volt_table[i].leakage_level) {
+			maxvolt = lkg_volt_table[i].maxvolt;
+			break;
+		}
+	}
+
+	maxvolt = maxvolt ? maxvolt : lkg_volt_table[i-1].maxvolt;
+
+	// limit high voltage
+	for (i = 0; table[i].frequency != CPUFREQ_TABLE_END; i++) {
+		if (table[i].index > maxvolt) {
+			printk("\t\tadjust table freq=%d KHz, index=%d mV", table[i].frequency, table[i].index);
+			table[i].index = maxvolt;
+			printk(" to index=%d mV\n", table[i].index);
+		}
+	}
+#if 0
+	/*
+	 * Low freq add some voltage for low leakage chip.
+	 * Open it when necessary.
+	 *
+	 */
+
+	// limit low voltage
+	if (strncmp(clk->dvfs_info->name, "cpu", strlen("cpu")) == 0
+			&& leakage_level <= LOW_LEAKAGE_BOUND) {
+		for (i = 0; table[i].frequency != CPUFREQ_TABLE_END; i++) {
+			if (table[i].index + VOLT_COMPENSATION < maxvolt) {
+				printk("\t\tadjust table freq=%d KHz, index=%d mV",
+						table[i].frequency, table[i].index);
+				table[i].index += VOLT_COMPENSATION;
+				printk(" to index=%d mV\n", table[i].index);
+			}
+		}
+	}
+#endif
+}
+
 #define NO_VOLT_DIFF
 #ifdef NO_VOLT_DIFF
 
@@ -730,9 +843,6 @@
 
 
 /******************************rk30 avs**************************************************/
-
-
-
 static void __iomem *rk30_nandc_base=NULL;
 
 #define nandc_readl(offset)	readl_relaxed(rk30_nandc_base + offset)
@@ -757,6 +867,8 @@
 	nandc_writel(nanc_save_reg[0] | 0x1 << 14, 0);
 	nandc_writel(0x5, 0x130);
 
+	/* Just break lock status */
+	nandc_writel(0x1, 0x158);
 	nandc_writel(0x3, 0x158);
 	nandc_writel(0x21, 0x134);
 
