--- a/arch/arm/plat-rk/rk-sdmmc-wifi.c
+++ b/arch/arm/plat-rk/rk-sdmmc-wifi.c
@@ -16,9 +16,7 @@
 static int rk29sdk_wifi_status(struct device *dev);
 static int rk29sdk_wifi_status_register(void (*callback)(int card_presend, void *dev_id), void *dev_id);
 
-// Omegamoon >> Disabled
 #if defined(CONFIG_WIFI_COMBO_MODULE_CONTROL_FUNC)
-// Omegamoon << Disabled
 static int rk29sdk_wifi_mmc0_status(struct device *dev);
 static int rk29sdk_wifi_mmc0_status_register(void (*callback)(int card_presend, void *dev_id), void *dev_id);
 static int rk29sdk_wifi_mmc0_cd = 0;   /* wifi virtual 'card detect' status */
@@ -28,9 +26,7 @@
 int rk29sdk_wifi_power_state = 0;
 int rk29sdk_bt_power_state = 0;
 
-// Omegamoon >> Disabled
     #if defined(CONFIG_USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD)
-// Omegamoon <<
     	/////////////////////////////////////////////////////////////////////////////////////
 	    // set the gpio to develop wifi EVB if you select the macro of CONFIG_USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD
 	    #define USE_SDMMC_CONTROLLER_FOR_WIFI   0
@@ -90,12 +86,8 @@
         #define RK30SDK_WIFI_GPIO_GPS_LAN_IOMUX_FMUX        GPIO4D_TRACE_DATA14
         #endif // #if COMBO_MODULE_MT6620_CDT--#endif
         
-// Omegamoon >> Disabled
     #endif // #if defined(CONFIG_USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD)---#endif
-// Omegamoon << Disabled
-// Omegamoon >> Disabled
 #endif // #if defined(CONFIG_WIFI_COMBO_MODULE_CONTROL_FUNC) ---#endif
-// Omegamoon << Disabled
 
 static int rk29sdk_wifi_cd = 0;   /* wifi virtual 'card detect' status */
 static void (*wifi_status_cb)(int card_present, void *dev_id);
@@ -254,6 +246,14 @@
 
 
 #ifdef CONFIG_WIFI_CONTROL_FUNC
+#if defined(CONFIG_USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD)
+static int rk29sdk_wifi_mmc0_status(struct device *dev);
+static int rk29sdk_wifi_mmc0_status_register(void (*callback)(int card_presend, void *dev_id), void *dev_id);
+static int rk29sdk_wifi_mmc0_cd = 0;   /* wifi virtual 'card detect' status */
+static void (*wifi_mmc0_status_cb)(int card_present, void *dev_id);
+static void *wifi_mmc0_status_cb_devid;
+#endif
+
 #define PREALLOC_WLAN_SEC_NUM           4
 #define PREALLOC_WLAN_BUF_NUM           160
 #define PREALLOC_WLAN_SECTION_HEADER    24
@@ -330,6 +330,23 @@
         return -ENOMEM;
 }
 
+
+#if defined(CONFIG_USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD)
+static int rk29sdk_wifi_mmc0_status(struct device *dev)
+
+{
+        return rk29sdk_wifi_mmc0_cd;
+}
+
+static int rk29sdk_wifi_mmc0_status_register(void (*callback)(int card_present, void *dev_id), void *dev_id)
+{
+        if(wifi_mmc0_status_cb)
+                return -EAGAIN;
+        wifi_mmc0_status_cb = callback;
+        wifi_mmc0_status_cb_devid = dev_id;
+        return 0;
+}
+#else
 static int rk29sdk_wifi_status(struct device *dev)
 {
         return rk29sdk_wifi_cd;
@@ -343,18 +360,26 @@
         wifi_status_cb_devid = dev_id;
         return 0;
 }
+#endif
 
 static int __init rk29sdk_wifi_bt_gpio_control_init(void)
 {
     rk29sdk_init_wifi_mem();    
+ #if !(!!SDMMC_USE_NEW_IOMUX_API)
     rk29_mux_api_set(rk_platform_wifi_gpio.power_n.iomux.name, rk_platform_wifi_gpio.power_n.iomux.fgpio);
+ #endif   
 
+#ifdef CONFIG_MACH_RK_FAC
+	if(wifi_pwr!=-1)
+		port_output_init(wifi_pwr, 1, "wifi_pwr"); 
+#else
     if (rk_platform_wifi_gpio.power_n.io != INVALID_GPIO) {
         if (gpio_request(rk_platform_wifi_gpio.power_n.io, "wifi_power")) {
                pr_info("%s: request wifi power gpio failed\n", __func__);
                return -1;
         }
     }
+#endif
 
 #ifdef RK30SDK_WIFI_GPIO_RESET_N
     if (rk_platform_wifi_gpio.reset_n.io != INVALID_GPIO) {
@@ -366,8 +391,13 @@
     }
 #endif    
 
-    if (rk_platform_wifi_gpio.power_n.io != INVALID_GPIO)
+#ifdef CONFIG_MACH_RK_FAC
+	if(wifi_pwr!=-1)
+		port_output_off(wifi_pwr);
+#else
+	if (rk_platform_wifi_gpio.power_n.io != INVALID_GPIO)
         gpio_direction_output(rk_platform_wifi_gpio.power_n.io, !(rk_platform_wifi_gpio.power_n.enable) );
+#endif
 
 #ifdef RK30SDK_WIFI_GPIO_RESET_N 
     if (rk_platform_wifi_gpio.reset_n.io != INVALID_GPIO)
@@ -403,22 +433,33 @@
     return 0;
 }
 
-#if (defined(CONFIG_RTL8192CU) || defined(CONFIG_RTL8188EU) )&& defined(CONFIG_ARCH_RK2928)
+#if (defined(CONFIG_RTL8192CU) || defined(CONFIG_RTL8188EU) || defined(CONFIG_RTL8723AU) || defined(CONFIG_RTL8192DU)) \
+	&& (defined(CONFIG_ARCH_RK2928) || defined(CONFIG_MACH_RK3026_86V) ||defined(CONFIG_MACH_RK3026_86V_FAC))
 static int usbwifi_power_status = 1;
 int rk29sdk_wifi_power(int on)
 {
         pr_info("%s: %d\n", __func__, on);
          if (on){
+            #if defined(CONFIG_USB_WIFI_POWER_CONTROLED_BY_GPIO)
+                gpio_set_value(rk_platform_wifi_gpio.power_n.io, rk_platform_wifi_gpio.power_n.enable);
+                mdelay(100);
+            #else
                 if(usbwifi_power_status == 1) {
                     rkusb_wifi_power(0);
                     mdelay(50);
                 }
                 rkusb_wifi_power(1);
+            #endif
                 usbwifi_power_status = 1;
                  pr_info("wifi turn on power\n");  	
         }else{
+            #if defined(CONFIG_USB_WIFI_POWER_CONTROLED_BY_GPIO)
+                gpio_set_value(rk_platform_wifi_gpio.power_n.io, !(rk_platform_wifi_gpio.power_n.enable));
+                mdelay(100);
+            #else
                 rkusb_wifi_power(0);
                 usbwifi_power_status = 0;    	
+            #endif
                  pr_info("wifi shut off power\n");
         }
         return 0;
@@ -428,7 +469,12 @@
 {
         pr_info("%s: %d\n", __func__, on);
         if (on){
+			#ifdef CONFIG_MACH_RK_FAC
+				if(wifi_pwr!=-1)
+					port_output_on(wifi_pwr);
+			#else
                 gpio_set_value(rk_platform_wifi_gpio.power_n.io, rk_platform_wifi_gpio.power_n.enable);
+			#endif
                 mdelay(50);
 
                 #if defined(CONFIG_SDMMC1_RK29) && !defined(CONFIG_SDMMC_RK29_OLD)	
@@ -443,8 +489,12 @@
                 pr_info("wifi turn on power\n");
         }else{
 //                if (!rk29sdk_bt_power_state){
-                        gpio_set_value(rk_platform_wifi_gpio.power_n.io, !(rk_platform_wifi_gpio.power_n.enable));
-
+				#ifdef CONFIG_MACH_RK_FAC
+						if(wifi_pwr!=-1)
+							port_output_off(wifi_pwr);
+				#else
+						gpio_set_value(rk_platform_wifi_gpio.power_n.io, !(rk_platform_wifi_gpio.power_n.enable));
+				#endif
                         #if defined(CONFIG_SDMMC1_RK29) && !defined(CONFIG_SDMMC_RK29_OLD)	
                         rk29_sdmmc_gpio_open(1, 0); //added by xbw at 2011-10-13
                         #endif
@@ -476,6 +526,19 @@
         return 0;
 }
 
+#if defined(CONFIG_USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD)
+int rk29sdk_wifi_set_carddetect(int val)
+{
+    printk("%s:%d\n", __func__, val);
+    rk29sdk_wifi_mmc0_cd = val;
+    if (wifi_mmc0_status_cb){
+            wifi_mmc0_status_cb(val, wifi_mmc0_status_cb_devid);
+    }else {
+            pr_warning("%s,in mmc0 nobody to notify\n", __func__);
+    }
+    return 0; 
+}
+#else
 int rk29sdk_wifi_set_carddetect(int val)
 {
         pr_info("%s:%d\n", __func__, val);
@@ -487,14 +550,62 @@
         }
         return 0;
 }
+#endif
 EXPORT_SYMBOL(rk29sdk_wifi_set_carddetect);
 
-#define WIFI_HOST_WAKE RK30_PIN3_PD2
+#include <linux/etherdevice.h>
+u8 wifi_custom_mac_addr[6] = {0,0,0,0,0,0};
+extern char GetSNSectorInfo(char * pbuf);
+int rk29sdk_wifi_mac_addr(unsigned char *buf)
+{
+       char mac_buf[20] = {0};
+    printk("rk29sdk_wifi_mac_addr.\n");
+
+    // from vflash
+    if(is_zero_ether_addr(wifi_custom_mac_addr)) {
+       int i;
+       char *tempBuf = kmalloc(512, GFP_KERNEL);
+       if(tempBuf) {
+           GetSNSectorInfo(tempBuf);
+           for (i = 506; i <= 511; i++)
+               wifi_custom_mac_addr[i-506] = tempBuf[i];
+           kfree(tempBuf);
+       } else {
+           return -1;
+       }
+    }
+
+	sprintf(mac_buf,"%02x:%02x:%02x:%02x:%02x:%02x",wifi_custom_mac_addr[0],wifi_custom_mac_addr[1],
+	wifi_custom_mac_addr[2],wifi_custom_mac_addr[3],wifi_custom_mac_addr[4],wifi_custom_mac_addr[5]);
+	printk("falsh wifi_custom_mac_addr=[%s]\n", mac_buf);
+
+	if (is_valid_ether_addr(wifi_custom_mac_addr)) {
+		if (2 == (wifi_custom_mac_addr[0] & 0x0F)) {
+			printk("This mac address come into conflict with the address of direct, ignored...\n");
+			return -1;
+		}
+	} else {
+		printk("This mac address is not valid, ignored...\n");
+		return -1;
+	}
+
+#if defined(CONFIG_RKWIFI)
+	memcpy(buf, wifi_custom_mac_addr, 6);
+#else 
+	memcpy(buf, mac_buf, strlen(mac_buf));//realtek's wifi use this branch
+#endif
+    return 0;
+}
+EXPORT_SYMBOL(rk29sdk_wifi_mac_addr);
+
+//#define WIFI_HOST_WAKE RK30_PIN3_PD2
 
 static struct resource resources[] = {
 	{
-		.start = WIFI_HOST_WAKE,
-		.flags = IORESOURCE_IRQ,
+#ifdef RK30SDK_WIFI_GPIO_WIFI_INT_B
+		.start = RK30SDK_WIFI_GPIO_WIFI_INT_B,
+#endif
+		.flags = IORESOURCE_IRQ | IORESOURCE_IRQ_HIGHLEVEL | IORESOURCE_IRQ_SHAREABLE,
 		.name = "bcmdhd_wlan_irq",
 	},
 };
@@ -519,9 +630,65 @@
 EXPORT_SYMBOL(rk29sdk_wifi_combo_get_GPS_SYNC_gpio);
 
 
+#if defined(CONFIG_MTK_COMBO_MT66XX)
+    static struct mtk_wmt_platform_data mtk_wmt_pdata = {
+        .pmu = RK30SDK_WIFI_GPIO_POWER_N,//RK30_PIN0_PB5, //MUST set to pin num in target system
+        .rst = RK30SDK_WIFI_GPIO_RESET_N,//RK30_PIN3_PD0, //MUST set to pin num in target system
+        .bgf_int = RK30SDK_WIFI_GPIO_BGF_INT_B,//RK30_PIN0_PA5,//MUST set to pin num in target system if use UART interface.
+        .urt_cts = -EINVAL, // set it to the correct GPIO num if use common SDIO, otherwise set it to -EINVAL.
+        .rtc = -EINVAL, //Optipnal. refer to HW design.
+        .gps_sync = -EINVAL, //Optional. refer to HW design.
+        .gps_lna = -EINVAL, //Optional. refer to HW design.
+    };
+    static struct mtk_sdio_eint_platform_data mtk_sdio_eint_pdata = {
+        .sdio_eint = RK30SDK_WIFI_GPIO_WIFI_INT_B,//53, //MUST set pin num in target system.
+    };
+    static struct platform_device mtk_wmt_dev = {
+        .name = "mtk_wmt",
+        .id = 1,
+        .dev = {
+       
+       
+        .platform_data = &mtk_wmt_pdata,
+        },
+    };
+    static struct platform_device mtk_sdio_eint_dev = {
+        .name = "mtk_sdio_eint",
+        .id = 1,
+        .dev = {
+        .platform_data = &mtk_sdio_eint_pdata,
+        },
+    };
+    static void __init mtk_combo_init(void)
+    {
+        /* gpio number align target system¡¯s setting */
+        gpio_request(mtk_wmt_pdata.pmu, "MT66XX PMUEN");
+        gpio_request(mtk_wmt_pdata.rst, "MT66XX SYSRST");
+        gpio_direction_output(mtk_wmt_pdata.pmu, 0);
+        gpio_direction_output(mtk_wmt_pdata.rst, 0);
+        gpio_free(mtk_wmt_pdata.pmu);
+        gpio_free(mtk_wmt_pdata.rst);
+        
+        /* step3. if use UART interface, config UART_CTS(host end) to UART_CTS mode;
+        if use common SDIO interface, config UART_CTS(host_end) to GPIO mode. */
+        if (mtk_wmt_pdata.urt_cts == -EINVAL) {
+            //UART interface,config to UART CTS mode
+        } else {
+            //SDIO interface,config to GPIO mode.
+            //omap_mux_set_gpio(3,mtk_wmt_pdata.urt_cts);
+        }
+        return;
+    }
+#endif---//#if defined(CONFIG_MTK_COMBO_MT66XX)
+
 static int rk29sdk_wifi_combo_module_gpio_init(void)
 {
-    //VDDIO
+#if defined(CONFIG_MTK_COMBO_MT66XX)
+    mtk_combo_init();
+    platform_device_register(&mtk_wmt_dev);
+    platform_device_register(&mtk_sdio_eint_dev);
+#else  
+  //VDDIO
     #ifdef RK30SDK_WIFI_GPIO_VCCIO_WL
         #ifdef RK30SDK_WIFI_GPIO_VCCIO_WL_PIN_NAME
         rk30_mux_api_set(rk_platform_wifi_gpio.vddio.iomux.name, rk_platform_wifi_gpio.vddio.iomux.fgpio);
@@ -590,6 +757,7 @@
 
 	#endif//#if COMBO_MODULE_MT6620_CDT ---#endif 
 
+#endif
     return 0;
 }
 
@@ -783,6 +951,7 @@
         .set_reset = rk29sdk_wifi_reset,
         .set_carddetect = rk29sdk_wifi_set_carddetect,
         .mem_prealloc   = rk29sdk_mem_prealloc,
+        .get_mac_addr   = rk29sdk_wifi_mac_addr,
 };
 
 static struct platform_device rk29sdk_wifi_device = {
