--- a/drivers/net/usb/Kconfig
+++ b/drivers/net/usb/Kconfig
@@ -241,6 +241,15 @@
 	  This option adds support for Davicom SR9700 based USB 2.0
 	  10/100 Ethernet adapters.
 
+config USB_NET_SR9800
+	tristate "WilLing Electrnic SR9800 based USB 2.0"
+	depends on USB_USBNET
+	select CRC32
+	default y
+	help
+	  This option adds support for Davicom SR9800 based USB 2.0
+	  10/100 Ethernet adapters.
+
 config USB_NET_SMSC75XX
 	tristate "SMSC LAN75XX based USB 2.0 gigabit ethernet devices"
 	depends on USB_USBNET
