--- a/drivers/power/power_supply_core.c
+++ b/drivers/power/power_supply_core.c
@@ -19,9 +19,6 @@
 #include <linux/power_supply.h>
 #include "power_supply.h"
 
-#ifdef CONFIG_POWER_ON_CHARGER_DISPLAY 
-extern struct list_head rk_psy_head;
-#endif
 /* exported for the APM Power driver, APM emulation */
 struct class *power_supply_class;
 EXPORT_SYMBOL_GPL(power_supply_class);
@@ -184,10 +181,6 @@
 
 	device_initialize(dev);
 
-#ifdef CONFIG_POWER_ON_CHARGER_DISPLAY 
-	list_add(&psy->rk_psy_node, &rk_psy_head);
-#endif
-
 	dev->class = power_supply_class;
 	dev->type = &power_supply_dev_type;
 	dev->parent = parent;
