--- a/include/linux/display-sys.h
+++ b/include/linux/display-sys.h
@@ -1,6 +1,5 @@
 #ifndef _LINUX_DISPLAY_RK_H
 #define _LINUX_DISPLAY_RK_H
-
 #include <linux/device.h>
 #include <linux/fb.h>
 #include <linux/list.h>
@@ -20,6 +19,25 @@
 	DISPLAY_SCALE_Y
 };
 
+enum rk_display_property {
+	DISPLAY_MAIN = 0,
+	DISPLAY_AUX
+};
+
+enum rk_display_source {
+	DISPLAY_SOURCE_LCDC0 = 0,
+	DISPLAY_SOURCE_LCDC1
+};
+
+/* HDMI mode list*/
+struct display_modelist {
+	struct list_head 	list;
+	struct fb_videomode	mode;
+	unsigned int 		vic;
+	unsigned int		format_3d;
+	unsigned int		detail_3d;
+};
+
 /* This structure defines all the properties of a Display. */
 struct rk_display_driver {
 	void (*suspend)(struct rk_display_device *, pm_message_t state);
@@ -37,6 +55,9 @@
 	int (*getmode)(struct rk_display_device *, struct fb_videomode *mode);
 	int (*setscale)(struct rk_display_device *, int, int);
 	int (*getscale)(struct rk_display_device *, int);
+	int (*get3dmode)(struct rk_display_device *);
+	int (*set3dmode)(struct rk_display_device *, int);
+	int (*getedidaudioinfo)(struct rk_display_device *, char *audioinfo, int len);
 };
 
 struct rk_display_device {
@@ -51,6 +72,7 @@
 	int idx;
 	struct rk_display_ops *ops;
 	int priority;
+	int property;
 	struct list_head list;
 };
 
@@ -59,6 +81,16 @@
 	struct rk_display_device *dev;
 };
 
+struct rkdisplay_platform_data {
+	int property;			//display screen property: main display or aux display.
+	int video_source;		//display screen video source
+	int io_pwr_pin;			//power control gpio
+	int io_reset_pin;		//reset control gpio
+	int io_switch_pin;		//cvbs/ypbpr output switch gpio
+};
+
+extern int display_add_videomode(const struct fb_videomode *mode, struct list_head *head);
+
 extern struct rk_display_device *rk_display_device_register(struct rk_display_driver *driver,
 					struct device *dev, void *devdata);
 extern void rk_display_device_unregister(struct rk_display_device *dev);
@@ -69,7 +101,7 @@
 extern void rk_display_device_disable_other(struct rk_display_device *ddev);
 
 
-extern void rk_display_device_select(int priority);
+extern void rk_display_device_select(int property, int priority);
 
 #define to_rk_display_device(obj) container_of(obj, struct rk_display_device, class_dev)
 
