--- a/drivers/mmc/host/Kconfig
+++ b/drivers/mmc/host/Kconfig
@@ -4,6 +4,18 @@
 
 comment "MMC/SD/SDIO Host Controller Drivers"
 
+config EMMC_RK
+	tristate "RK emmc controller suppport"
+	depends on (!SDMMC2_RK29)&&(ARCH_RK3188 || ARCH_RK3066B || ARCH_RK3026 || ARCH_RK319X)
+	default y
+	help
+		This selects the RK EMMC controller
+if EMMC_RK
+	config EMMC_DDR_MODE
+		default y
+		bool "RK emmc ddr mode support"
+endif
+
 config SDMMC_RK29
 	tristate "RK29 SDMMC controller suppport"
 	depends on PLAT_RK
@@ -42,13 +54,13 @@
 
 			   If you say No, then detect the card by register interrupt.
 
-#	config USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD
-#	    depends on SDMMC0_RK29
-#		bool "Switch the driver SDMMC0 for the debug of wifi_develop_board."
-#		default n
-#		help
-#			In order to debug the Wifi development board using SD interface,
-#			we can switch the driver SDMMC0.
+	config USE_SDMMC0_FOR_WIFI_DEVELOP_BOARD
+	    depends on SDMMC0_RK29
+		bool "Switch the driver SDMMC0 for the debug of wifi_develop_board."
+		default n
+		help
+			In order to debug the Wifi development board using SD interface,
+			we can switch the driver SDMMC0.
 
 
 	config SDMMC1_RK29
@@ -67,20 +79,36 @@
        config  RK29_SDIO_IRQ_FROM_GPIO
          bool "sdio-irq from gpio"
          default n
-         depends on SDMMC1_RK29
+         depends on SDMMC1_RK29 && (BCM_OOB_ENABLED || MT6620)
          help
             You will generate sdio interrupt from gpio if you say Yes.
             Please note that this feature requires hardware support.
 
 
 
-#	config USE_SDMMC1_FOR_WIFI_DEVELOP_BOARD
-#	  depends on SDMMC1_RK29
-#		bool "Switch the driver SDMMC1 for the debug of wifi_develop_board."
-#		default n
-#		help
-#			In order to debug the Wifi development board using SD interface,
-#			we can switch the driver SDMMC1.
+	config USE_SDMMC1_FOR_WIFI_DEVELOP_BOARD
+	  depends on SDMMC1_RK29
+		bool "Switch the driver SDMMC1 for the debug of wifi_develop_board."
+		default n
+		help
+			In order to debug the Wifi development board using SD interface,
+			we can switch the driver SDMMC1.
+
+	config SDMMC1_RK29_SDCARD_DET_FROM_GPIO
+	   depends on SDMMC1_RK29 && USE_SDMMC1_FOR_WIFI_DEVELOP_BOARD
+	    bool "use the gpio-interrupt to detect mmc1-card"
+		default n
+		help
+		    You can detect the presence of card by the gpio-interrupt if you say Yes.
+			Of course, you must define the pin-name for the detect-pin.
+			If you say No, then detect the card by register interrupt.
+
+        config SDMMC2_RK29
+                tristate "RK29 SDMMC2 controller support(eMMC)"
+                default n
+                help
+                        This supports the use of the SDMMC2 controller on Rk processors.
+
 
 endif
 
